core/src/main/java/fr/sncf/osrd/api/PathfindingResult.java

```diff
@@ -30,30 +32,77 @@ public class PathfindingResult {
+     *
      * @param path contains the pathfinding's result
      */
     public static PathfindingResult make(
-            List<Pathfinding.EdgeRange<SignalingRoute>> path,
+            Pathfinding.Result<SignalingRoute> path,
             SignalingInfra infra,
             WarningRecorderImpl warningRecorder
     ) {
         var res = new PathfindingResult();
-        for (var signalingRouteEdgeRange : path)
+        for (var signalingRouteEdgeRange : path.ranges())
             if (signalingRouteEdgeRange.start() < signalingRouteEdgeRange.end())
                 res.routePaths.add(
                         makeRouteResult(
                                 res,
                                 signalingRouteEdgeRange
                         )
                 );
-        var lastRoute = res.routePaths.get(res.routePaths.size() - 1);
-        var lastRange = lastRoute.trackSections.get(lastRoute.trackSections.size() - 1);
-        res.addStep(new PathWaypointResult(lastRange.trackSection.id.id, lastRange.end));
         res.addGeometry(infra);
         res.warnings = warningRecorder.warnings;
+        setAllWaypoints(res, path);
         return res;
     }
 
+    /** Adds all missing waypoints and sorts them */
+    private static void setAllWaypoints(PathfindingResult res, Pathfinding.Result<SignalingRoute> path) {
+        for (var waypoint : path.waypoints())
+            res.addStep(convertLocation(waypoint));
+        var waypointsPerTrack = HashMultimap.<String, PathWaypointResult>create();
+        for (var waypoint : res.pathWaypoints)
+            waypointsPerTrack.put(waypoint.track.id.id, waypoint);
+        double routeOffset = 0;
+        for (var routeRange : path.ranges()) {
+            double trackOffset = routeOffset - routeRange.start();
+            for (var trackRange : routeRange.edge().getInfraRoute().getTrackRanges()) {
+                for (var waypoint : waypointsPerTrack.get(trackRange.track.getEdge().getID())) {
+                    if (trackRange.containsOffset(waypoint.position)) {
+                        waypoint.pathOffset = trackOffset + trackRange.offsetOf(waypoint.position);
+                    }
+                }
+                trackOffset += trackRange.getLength();
+            }
+            routeOffset += routeRange.end() - routeRange.start();
+        }
+        for (var waypoint : res.pathWaypoints)
+            assert !Double.isNaN(waypoint.pathOffset) : "Waypoint couldn't be found in the path";
+        res.pathWaypoints.sort(Comparator.comparingDouble(w -> w.pathOffset));
+
+        // Remove duplicates
+        for (int i = 1; i < res.pathWaypoints.size(); i++) {
+            if (res.pathWaypoints.get(i - 1).isDuplicate(res.pathWaypoints.get(i))) {
+                res.pathWaypoints.get(i - 1).merge(res.pathWaypoints.get(i));
+                res.pathWaypoints.remove(i);
+                i--;
+            }
+
+        }
```

Comment:
if my above suggestion works, this is not needed anymore, as addstep already deduplicates